designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+        observedGeneration --> reconcile: no
+        observedGeneration --> [*]: yes
+    delete --> done
+    reconcile --> done
+    state done <<choice>>
+        done --> updateGeneration: yes
+        done --> Queue: no
+    updateGeneration  --> [*]
+```
+
+We only perform the actual reconciliation logic (this is trying to change the state of the world) when `observedGeneration` is different than `generation`. This will only happen after the resource has changed.
+
+It's important to mention that we only update the value of `observedGeneration` when the reconciliation process is "done". Done here means that the state of the world is equal to the desired state: the cluster has been completely reconciled. This is to take into account the case where a full reconciliation takes multiple loops (for example, when the cluster is first created). Note that "reconcile not done" also includes errors.
+
+This might be slightly different of how other resources use the `observedGeneration` field, where the field is always set on every reconciliation loop. For multi state resources (like our `Cluster`) this is usually accompanied of a "state" indicator (note that `phases` is now discouraged in favor of `conditions`). However, we don't have that in our `Cluster`'s status. Although tempting to try to add this to the status, we encourage to tackle this problem separately and revisit the use of `observedGeneration` if necessary. Refer to the [alternatives](#alternative-solutions) section for other possible solutions.
+
+### Cluster vs cluster config: how to calculate an aggregated generation
+An astute reader might have realized that until now we have worked with a simplified model: where cluster reconciliation only needs to occur when the resource been reconciled `Cluster` object changes. Although this is true for most controllers in Kubernetes, our Cluster controller doesn't behave this way. Due to the particularities of our API and how we define a single entity (the cluster config) with multiple CRDs (all linked through the `Cluster` object), the Cluster controller reacts to changes not only in the `Cluster` but also in all its linked objects (we'll call them child objects for the rest of the document).
+
+Now, the question becomes: how do we reflect child objects changes in the `observedGeneration`? Since all our CRDs, including the child objects, have a `generation` field, we can simply add all of them into one aggregated cluster config `observed`. The are two different ways for a referenced object to change:
+* If a child object reference is changed to a different object, `Cluster` object's `generation` will change, increasing the aggregated cluster config's generation.
+* If the reference is kept pointing to the same object, but the child object is changed, this will increase its `generation`, which also increases the aggregated cluster config's generation.
+$$aggregatedGeneration = \sum_{n=1}^{clusterConfigObjs} generation_n$$
```

Comment:
Are you saying this a new field or using the same `observedGeneration` field in cluster config?

Reply:
At this point this is talking about setting `observedGeneration` with this calculation.